javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1014,6 +1043,201 @@ object Formatter {
+   *  using its built-in `toString()` conversion. By ECMAScript's spec, this
+   *  yields a decimal representation with as few significant digits as
+   *  possible, although it can be in fixed notation or in computerized
+   *  scientific notation.
+   *
+   *  We then parse that string to recover the integer part, the factional part
+   *  and the exponent; the latter two being optional.
+   *
+   *  From the parts, we construct a `Decimal`, making sure to create one that
+   *  does not have leading 0's (as it is forbidden by `Decimal`'s invariants).
+   */
+  private def numberToDecimal(x: Double): Decimal = {
+    if (x == 0.0) {
+      val negative = 1.0 / x < 0.0
+      Decimal.zero(negative)
+    } else {
+      val negative = x < 0.0
+      val s = JDouble.toString(if (negative) -x else x)
+
+      val ePos = s.indexOf('e')
+      val e =
+        if (ePos < 0) 0
+        else js.Dynamic.global.parseInt(s.substring(ePos + 1)).asInstanceOf[Int]
+      val significandEnd = if (ePos < 0) s.length() else ePos
+
+      val dotPos = s.indexOf('.')
+      if (dotPos < 0) {
+        // No '.'; there cannot be leading 0's (x == 0.0 was handled before)
+        val unscaledValue = s.substring(0, significandEnd)
+        val scale = -e
+        new Decimal(negative, unscaledValue, scale)
+      } else {
+        // There is a '.'; there can be leading 0's, which we must remove
+        val digits = s.substring(0, dotPos) + s.substring(dotPos + 1, significandEnd)
+        val digitsLen = digits.length()
+        var i = 0
+        while (i < digitsLen && digits.charAt(i) == '0')
+          i += 1
+        val unscaledValue = digits.substring(i)
+        val scale = -e + (significandEnd - (dotPos + 1))
+        new Decimal(negative, unscaledValue, scale)
+      }
+    }
+  }
+
+  /** A decimal representation of a number.
+   *
+   *  An instance of this class represents the number whose absolute value is
+   *  `unscaledValue × 10^(-scale)`, and that is negative iff `negative` is
+   *  true.
+   *
+   *  The `unscaledValue` is stored as a String of decimal digits, i.e.,
+   *  characters in the range ['0', '9'], expressed in base 10. Leading 0's are
+   *  *not* valid.
+   *
+   *  As an exception, a zero value is represented by an `unscaledValue` of
+   *  `"0"`. The scale of zero value is always 0.
+   *
+   *  `Decimal` is similar to `BigDecimal`, with some differences:
+   *
+   *  - `Decimal` distinguishes +0 from -0.
+   *  - The unscaled value of `Decimal` is stored in base 10.
+   *
+   *  The methods it exposes have the same meaning as for BigDecimal, with the
+   *  only rounding mode being HALF_UP.
+   */
+  private final class Decimal(val negative: Boolean, val unscaledValue: String,
+      val scale: Int) {
+
+    def isZero: Boolean = unscaledValue == "0"
+
+    /** The number of digits in the unscaled value.
+     *
+     *  The precision of a zero value is 1.
+     */
+    def precision: Int = unscaledValue.length()
```

Comment:
Instead of exposing `precision` here, consider exposing `orderOfMagnitude`. This is the only usage, but it is a value that does not depend on the internal representation (scale v.s. trailing zeroes).

Reply:
I wanted to stick to the concepts that `BigDecimal` defines. I can expose `orderOfMagnitude`, but I still need to talk about the concept of `precision` to specify `round`, for example. So I'm not sure hiding it is going to help.

Reply:
Yeah, that's fair. Maybe consider simply not defining this helper and use `unscaledValue.length()` at the call site (not 100% sure this is better).

Reply:
Hopefully, now that I've renamed the conflicting uses of "precision", this shouldn't be an issue anymore.